src/services/membership/membership.service.ts

```diff
@@ -0,0 +1,108 @@
+import { Inject, LoggerService } from '@nestjs/common';
+import { WINSTON_MODULE_NEST_PROVIDER } from 'nest-winston';
+import { EcoverseService } from '@domain/challenge/ecoverse/ecoverse.service';
+import { OrganisationService } from '@domain/community/organisation/organisation.service';
+import { Membership } from './membership.dto.result';
+import { UserService } from '@domain/community/user/user.service';
+import { MembershipInput } from './membership.dto.input';
+import { MembershipEcoverseResultEntry } from './membership.dto.result.ecoverse.entry';
+import { AuthorizationCredential } from '@core/authorization';
+import { CommunityService } from '@domain/community/community/community.service';
+import { Community } from '@domain/community/community';
+import { UserGroupService } from '@domain/community/user-group/user-group.service';
+import { IChallenge } from '@domain/challenge/challenge';
+import { IUserGroup } from '@domain/community/user-group';
+import { MembershipResultEntry } from './membership.dto.result.entry';
+
+export class MembershipService {
+  constructor(
+    private userService: UserService,
+    private userGroupService: UserGroupService,
+    private communityService: CommunityService,
+    private ecoverseService: EcoverseService,
+    private organisationService: OrganisationService,
+    @Inject(WINSTON_MODULE_NEST_PROVIDER) private readonly logger: LoggerService
+  ) {}
+
+  async membership(membershipData: MembershipInput): Promise<Membership> {
+    const membership = new Membership();
+    const user = await this.userService.getUserWithAgent(membershipData.userID);
+    const credentials = user.agent?.credentials;
+    if (!credentials) {
+      return membership;
+    }
+    const storedChallenges: IChallenge[] = [];
+    const storedUserGroups: IUserGroup[] = [];
+    for (const credential of credentials) {
```

Comment:
Design-wise, we should consider adding a Membership name (I don't have a good name for it) that is implemented by the services and has a method that receives a resourceID. A lot of the code below will be simplified and instead of getting an organization and then resolving its membership, we can resolve the membership by passing the resourceID.

Reply:
agreed that this can be more efficient. 
already realised when doing it that we need to carefully look again at the set of credentials e.g. community member is not right, it should be ecoverse member + challenge member. 
Also ties into the notion of authorisation scopes that we talked about last week.

So not pick up here, but have added a note to https://app.zenhub.com/workspaces/cherrytwist-5ecb98b262ebd9f4aec4194c/issues/cherrytwist/cherrytwist/234 so that we pick it up with that phase 2 authorisation work.

Reply:
I see nothing in this epic reflecting my suggestion. Agreed that we might postpone some fixes to be done in the grand schema of things but we should at the very least capture our thoughts. Let's chat this through.